test/FilenameUtilsTest.cpp

```diff
@@ -0,0 +1,510 @@
+#include "pch.h"
+#include <kf/FilenameUtils.h>
+
+using namespace kf;
+
+SCENARIO("FilenameUtils getPathNoEndSeparator")
+{
+    GIVEN("various file paths")
+    {
+        WHEN("getting path from empty path")
+        {
+            USimpleString emptyPath(L"");
+            auto result = FilenameUtils::getPathNoEndSeparator(emptyPath);
+
+            THEN("result is empty")
+            {
+                REQUIRE(result.isEmpty());
+            }
+        }
+
+        WHEN("getting path from path with several trailing slashes")
+        {
+            USimpleString pathWithTrailingSlashes(L"\\Device\\HarddiskVolume1\\Windows\\\\\\");
+            auto result = FilenameUtils::getPathNoEndSeparator(pathWithTrailingSlashes);
+
+            THEN("path is extracted without trailing separator")
+            {
+                REQUIRE(result.equals(USimpleString(L"\\Device\\HarddiskVolume1\\Windows\\\\")));
+            }
+        }
+
+        WHEN("getting path from path that consists only of slashes")
+        {
+            USimpleString slashesOnly(L"\\\\\\");
+            auto result = FilenameUtils::getPathNoEndSeparator(slashesOnly);
+
+            THEN("result is path without last slash")
+            {
+                REQUIRE(result.equals(USimpleString(L"\\\\")));
+            }
+        }
+
+        WHEN("getting path from path without any slashes")
+        {
+            USimpleString noSlashes(L"filename.txt");
+            auto result = FilenameUtils::getPathNoEndSeparator(noSlashes);
+
+            THEN("result is empty")
+            {
+                REQUIRE(result.isEmpty());
```

Comment:
This is not what is really required from this function. Logically it should return output the same as input because there is just no trailing slash. So please fix the function itself. Then look forward for similar logical mistakes and fix both code and test. Report results.